src/main/scala/com/fulcrumgenomics/bam/ReadAndRefPosIterator.scala

```diff
@@ -179,3 +180,61 @@ object ReadAndRefPosIterator {
     )
   }
 }
+
+
+/** An iterator for each _mapped_ read base (i.e. has a corresponding reference base) that also has a _mapped_ base in
+  * its mate, with each value being the 1-based position in the read and reference respectively.
+  *
+  * Insertions and deletions will not be returned.  For example, an insertion consumes a read base, but has no
+  * corresponding reference base.  A deletion consumes a reference base, but has no corresponding read base.
+  *
+  * The _mapped_ read bases returned can be limited using the arguments for the minimum and maximum position in the read.
+  *
+  * @param rec the record over which read and reference positions should be returned
+  * @param mate the mate of the record
+  * @param minReadPos the minimum 1-based position in the read to return
+  * @param maxReadPos the maximum 1-based inclusive end position in the read to return
+  */
+class ReadMateAndRefPosIterator(rec: SamRecord,
+                                mate: SamRecord,
+                                minReadPos: Int = 1,
+                                maxReadPos: Int = Int.MaxValue) extends Iterator[ReadMateAndRefPos] {
+  // Use an iterators that returns the position for the read and mate respectively, then just find the reference positions
+  // where both have a read position defined.
+  private val recIter  = ReadAndRefPosIterator(rec=rec, mate=mate, minReadPos=minReadPos, maxReadPos=maxReadPos).buffered
+  private val mateIter = ReadAndRefPosIterator(rec=mate, mate=rec).buffered
+
+  private var nextItem: Option[ReadMateAndRefPos] = None
+
+  override def hasNext(): Boolean = {
+    while (nextItem.isEmpty && recIter.hasNext && mateIter.hasNext) {
+      val nextRec  = recIter.head
+      val nextMate = mateIter.head
+      if (nextRec.refPos < nextMate.refPos) recIter.next()
+      else if (nextMate.refPos < nextRec.refPos) mateIter.next()
+      else {
+        nextItem = Some(ReadMateAndRefPos(readPos=nextRec.readPos, matePos=nextMate.readPos, refPos=nextRec.refPos))
+        recIter.next()
+        mateIter.next()
+      }
+    }
+    nextItem.isDefined
+  }
+
+  def next(): ReadMateAndRefPos = this.nextItem match {
```

Comment:
I feel strongly that this should call/check `hasNext()`.  The way this is done, if I construct an iterator where I _know_ there should be 3 values in it, and then call `iter.next()` three times it will fail because it relies on `.hasNext()` to do the advancing, but nothing says you have to call `.hasNext()`.

I would re-write this as:

```scala
if (!hasNext()) throw new NoSuchElementException
val retval = this.nextItem.get
this.nextItem = None
retval
```

Reply:
done